designs/DeveloperTestingPackageExperience.md

```diff
@@ -0,0 +1,125 @@
+keep the design simple and to emphasise that it is not a replacement for other
+package sources. Also, since the target audience is creating the packages, the
+file would have been produced in order to publish it, so this limitation is not
+perceived as an impediment.
+
+## Solution Overview
+
+In order to provide a solution, there are a few issues that need to be
+highlighted with the current machinery with regards to installing of a package
+in the context of trying to test a newly created package:
+
+* NuGet caches packages by ID and version
+* The order of multiple sources are searched through is non-deterministic
+* If a version number is omitted, NuGet searches for the lowest applicable
+  version
+
+### `csproj` Additions
+
+The proposed solution is to add an additional MsBuild property for the
+[restore target](https://docs.microsoft.com/en-us/nuget/reference/msbuild-targets#restore-target)
+as follows:
+
+| Property              | Description  |
+|-----------------------|--------------|
+| RestoreOverrideFolder | Specifies a folder that contains `nupkg` files that take precedence over any other source and are not cached. |
+
+This property can only be specified once and can only specify a single folder
+(i.e. there cannot be multiple override folders to avoid ordering issues). Note
+that file shares _should_ be supported, however.
+
+Also, only a single version of a package (based on its ID) may exist in the
+override folder (i.e. it would be an error to have
+`generated.package.1.0.0.nupkg` and `generated.package.2.0.0.nupkg` in the
+folder).
+
+### Resolving Changes
+
+When an override folder is specified, any packages found inside the folder
+should be removed from the current cache (matching by ID and version) before any
+further work is done (i.e. this would delete the folder from the global cache
+or from the `PACKAGES_DIRECTORY` if specified in the
+[restore options](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-restore#options)).
+
+When a package that exists inside the overrides folder is referenced by the
+project:
+
+* If no version is specified, the version in the folder is used without
+  searching any other sources.
+* If a version is specified and that matches that in the override folder (taking
+  into account the usual wildcard support), that package is used without
+  searching any other sources.
+* Otherwise, normal resolution happens.
+
+When the override package is used, it is restored to the cache as normal (taking
+into account if `--no-cache` has been specified, a different cache folder used
+etc). This prevents changes to the rest of the restore process and doesn't
+affect being able to publish a different package of the same version to the
+override folder, as the version will be automatically cleaned during each
+restore (i.e. although it's installed in the cache, it will be removed from it
+again the next time the restore process starts).
```

Comment:
I think a `neater` approach instead of deleting packages would be to have *an additional* folder where these packages are consumed from. 

Deleting packages from the global packages folder automatically *can* be tricky because multiple projects share the global packages folder.

Reply:
Yep, that sounds good. I personally use a different folder to restore to (via the `--packages` option), so we could further limit the override to only work under that scenario if it helps?